sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,176 @@
+    cdef double* x_p
+    cdef int n_samples = X_.shape[0]
+    cdef int n_features = X_.shape[1]
+    cdef int point_index, center_index, label
+    cdef float upper_bound, distance
+    cdef double[:, :] center_distances = euclidean_distances(centers_) / 2.
+    cdef double[:, :] lower_bounds = np.zeros((n_samples, n_clusters))
+    cdef double[:] distance_next_center
+    labels_ = np.empty(n_samples, dtype=np.int32)
+    cdef int[:] labels = labels_
+    upper_bounds_ = np.empty(n_samples, dtype=np.float)
+    cdef double[:] upper_bounds = upper_bounds_
+    update_labels_distances_inplace(X_p, centers_p, center_distances, labels,
+                                    lower_bounds, upper_bounds, n_samples,
+                                    n_features, n_clusters)
+    cdef np.uint8_t[:] bounds_tight = np.ones(n_samples, dtype=np.uint8)
+    cdef np.uint8_t[:] points_to_update = np.zeros(n_samples, dtype=np.uint8)
+
+    if max_iter <= 0:
+        raise ValueError('Number of iterations should be a positive number'
+        ', got %d instead' % max_iter)
+
+    for iteration in range(max_iter):
+        if verbose:
+            print("start iteration")
+        # we could find the closest center in O(n) but
+        # this does not seem to be the bottleneck
+        distance_next_center = np.sort(center_distances, axis=0)[1]
+        if verbose:
+            print("done sorting")
+        for point_index in range(n_samples):
+            upper_bound = upper_bounds[point_index]
+            label = labels[point_index]
+            if distance_next_center[label] >= upper_bound:
+                continue
+            x_p = X_p + point_index * n_features
+            # check other update conditions
+            # TODO: get pointer to lower_bounds[point_index, center_index]
+            for center_index in range(n_clusters):
+                if (center_index != label
+                        and (upper_bound > lower_bounds[point_index, center_index])
+                        and (upper_bound > center_distances[center_index, label])):
+                    # update distance to center
+                    if not bounds_tight[point_index]:
+                        upper_bound = d(x_p, centers_p + label * n_features, n_features)
+                        lower_bounds[point_index, label] = upper_bound
+                        bounds_tight[point_index] = 1
+                    # check for relabels
+                    if (upper_bound > lower_bounds[point_index, center_index]
+                            or (upper_bound > center_distances[label, center_index])):
+                        distance = d(x_p, centers_p + center_index * n_features, n_features)
+                        lower_bounds[point_index, center_index] = distance
+                        if distance < upper_bound:
+                            label = center_index
+                            upper_bound = distance
+            labels[point_index] = label
```

Comment:
This and the below line should be in the inner `if distance < upper_bound` block, so that the labels are reassigned only if needed. Also label and upper_bound are unused below.

    if distance < upper_bound:
        labels[point_index] = center_index
       upper_bounds[point_index] = upper_bound